### PR TITLE
Fix reset password functionality

### PR DESCRIPTION
**Pull request by pli888**
# Pull request for issue: #859 

This is a pull request for the following functionalities:

* Re-implementation of reset password functionality with improved security based on good standard practices - see [here](https://paragonie.com/blog/2016/09/untangling-forget-me-knot-secure-account-recovery-made-simple). Essentially, the token consists of 2 parts: a selector and a verifier, and the token is only valid for use for 1 hour. The token validation process involves calculating the hash of the verifier and comparing it with the hash that has been saved in the database when the token was first created. If the hashes match then the token is valid if it has not expired.

The reset password functionality for website users consists of 2 parts:

1. The entry point is the display of the forgot password form at [`/site/forgot`](http://gigadb.gigasciencejournal.com:9170/site/forgot) which requires the user to enter in their email address. On pressing `Reset` then the `/site/thanks` page will be displayed.
2. An email will be sent to the entered email address which contains a link in the format of `/site/reset?token=selector1234verifierabcd`. Clicking this link containing a valid token will take you to the `/site/reset` page which asks you to input a new password. When this is successfully done then the web site will redirect you to the `/site/login` page. Password reset links with invalid tokens will take you to the `/site/forgot` page. There is an example [token URL](/site/reset?token=LcRiT6D70CBa1J9umjRMbgjhfE31Y3Bmd62qSvdm) that works in the local environment because of the data in `data/dev/reset_password_request.csv`. This link will only work once (as the token is deleted in the database) and then you will need to run `ops/scripts/setup_dev.sh` to use it again.

## Changes to the database schema

* `m300000_000000_drop_tables.php` has been updated to drop the `reset_password_request` table.
* The schema migration file,`m200528_052860_create_gigadb_user_tab.php` has been updated so that `gigadb_user_id_seq` will start providing `ids` in `gigadb_user` table from 400 which is a larger number than the largest id in the `gigadb_user.csv` migration script to populate the `gigadb_user` table with data.
* A new `reset_password_request` table has been added using `m200529_036000_create_reset_password_request.php`.

## Changes to the component classes

* A `CryptoService` component class has been created. It contains functions used to generate random strings and hashes from them.
* The new `PasswordResetTokenUserIdentity` class with its `authenticate` function is responsible for checking the validity of a password reset token that is sent to users by email within in an URL.

## Changes to the model classes

* The new `ForgotPasswordForm.php` provides a form model for users to enter their email address to receive the reset password email.
* A new `ResetPasswordForm` model is used for users to enter their new user account password.
* The `ResetPasswordRequest` model is used as the active record class for the `reset_password_request` table. If contains a `beforeSave()` function which will automatically create the `request_at` and `expires_at` timestamps if these are not provided to it.

## Changes to the controller classes

* The new `ResetPasswordRequestController` class provides the following action functions:
  * actionForgot() will display the Forgotten password page with a form for a user to enter their email address. It will also process the form, make checks on the user's email before sending a password reset email to them.
  * actionReset() will check the password reset token if it is valid and then display the reset password page.
  * actionThanks() is a simple action function for displaying the thanks page telling the user to check their email inbox for the password reset email.

* Various functions in `UserController` class have been deleted since their functionality has been replaced by `ResetPasswordRequestController`.

## Changes to the view classes in `protected/views/resetPasswordRequest` directory

* New `protected/views/resetPasswordRequest/emailReset.php` used for the HTML email send to users containing the reset password link. 
* New `protected/views/resetPasswordRequest/forgot.php` used for displaying Forgotten password form web page.
* New `protected/views/resetPasswordRequest/reset.php` used for displaying the Reset Password web page that users use to enter a new password.
* New `protected/views/resetPasswordRequest/thanks.php` used for displaying a web page to thank users for requesting their password to be reset.
* `protected/views/site/login.php` has been been updated so that the `Lost password` link goes to `site/forgot`.
* Various view classes used by the old password reset functionality have been deleted.

## Changes to the tests

* A new user record with `user@mailinator.com` email has been added into `gigadb_user.csv` and also a record in `reset_password_request.csv` that is an unexpired token to provide data for `LostPasswordPage.feature` acceptance test.
* The `features/lost-password-page.feature` Behat acceptance test has been replaced with the Codeception `LostPasswordPage.feature` acceptance test.
* A new `tests/acceptance/LostPasswordPage.feature` contains acceptance tests for the new password reset functionality. Note that it is does not test for the presence of the flash messages.
* `EmailCest` has been updated with a new `trySendPasswordEmail` test to check for the ability to send the password reset email to users. Can use command for run this test:
```
docker-compose run --rm test ./vendor/codeception/codeception/codecept run --debug tests/functional/EmailCest.php:^trySendPasswordEmail$
```
* New integration tests can be found in `ResetPasswordCest.php`. It contains tests for various reset token scenarios ensuring they are securely managed by checking for their presence in the `reset_password_request` database table. Can use command for run this test:
```
docker-compose run --rm test ./vendor/codeception/codeception/codecept run --debug tests/functional/ResetPasswordCest.php
```
* There is a unit `CryptoTest.php` test which perform some tests on the `CryptoService` component class.

## Changes to the provisioning

* `.gitlab-ci/yml` has been updated with `APPLICATION` attribute changed from `/builds/$CI_PROJECT_PATH` to `$CI_PROJECT_DIR`.
* `local.php.dist` has been updated to reference the `CryptoService` component and the new `HASH_SECRET_KEY` Gitlab variable. The `HASH_SECRET_KEY` has been added as Gitlab variable into the `Forks` group so it does not need to be created in your `GitlabID-gigadb-website` project variables.
* The urlManager section in `main.php.dist` has been updated with more aesthetic URL paths linking to the `ResetPasswordRequest` controller action functions.
* A new `triple-stash.migration.php.dist` file has been created which is used to generate the migration file for the reset_password_request database table. The reason for this is because the hash_token values can contain special characters which are escaped by Handlebars and we don't want this to happen that is achieved by using `{{{this}}}`.
* `csv_yii_migration.js` has been updated to process the`reset_password_request` table data using `triple-stash.migration.php.dist`.
* `generate_config.sh` has been updated to parameterise the $HASH_SECRET_KEY from Gitlab variables into `protected/config/local.php`.

### CODE REVIEW

**Review by rija — COMMENTED** [review]
Hi @pli888,

I noticed you have implemented from scratch a method to generate a random string in ``CryptoService``.

However Yii2 has a method for doing that. You can see it being used in a couple places in the codebase like:

* protected/tests/unit/DatasetUploadTest.php

```
		$writer = new Csv($spreadsheet);
		$testFile = sys_get_temp_dir().Yii::$app->security->generateRandomString(4).".csv" ;
		$writer->save($testFile);

```

* fuw/app/console/controllers/IdentityController.php

```
        $user->generateAuthKey();
        $user->generatePasswordResetToken();
        $user->setPassword(Yii::$app->security->generateRandomString(16));
        $user->save();

```

as well as in Yii2 framework itself.


If you look at the implementation of their method, it's also based on ``random_bytes()``. 

So it's probably a good idea to update ``CryptoService::getRandomAlphNumStr()`` to use the Yii2-provided method

The source code of the Yii2 method is in ``vendor/yiisoft/yii2/base/Security.php``, may be there are other useful methods in there?

**Comment by rija** [comment]
Hi @pli888,

When running the tests locally, I have failures in ``EmailCest.php`` if  I run the tests multiple time


For example:

```
$ docker-compose run --rm test ./vendor/codeception/codeception/codecept run -vvv functional EmailCest

```

Gives this output in from the second run onward:

```
...
There were 2 failures:

---------
1) EmailCest: Try send activation email
 Test  tests/functional/EmailCest.php:trySendActivationEmail
 Step  See in current url "/user/welcome"
 Fail  Failed asserting that '/user/create' contains "/user/welcome".

Scenario Steps:

 14. $I->seeInCurrentUrl("/user/welcome") at tests/functional/EmailCest.php:89
 13. $I->click("Register") at tests/functional/EmailCest.php:86
 12. $I->fillField({"id":"User_verifyCode"},"ouch") at tests/functional/EmailCest.php:85
 11. $I->checkOption("#User_terms") at tests/functional/EmailCest.php:84
 10. $I->checkOption("#User_newsletter") at tests/functional/EmailCest.php:83
 9. $I->selectOption("form select[id=User_preferred_link]","NCBI") at tests/functional/EmailCest.php:82

/var/www/vendor/phpunit/phpunit/src/Framework/Constraint.php:111
/var/www/vendor/phpunit/phpunit/src/Framework/Constraint.php:56
/var/www/vendor/phpunit/phpunit/src/Framework/Assert.php:2060
/var/www/vendor/phpunit/phpunit/src/Framework/Assert.php:146
/var/www/vendor/codeception/phpunit-wrapper/src/TestCase.php:47
/var/www/vendor/codeception/codeception/src/Codeception/Util/Shared/Asserts.php:449
/var/www/vendor/codeception/codeception/src/Codeception/Lib/InnerBrowser.php:562
/var/www/vendor/codeception/codeception/src/Codeception/Step.php:266
/var/www/vendor/codeception/codeception/src/Codeception/Scenario.php:72
/var/www/tests/_support/_generated/FunctionalTesterActions.php:1863
/var/www/tests/functional/EmailCest.php:89
/var/www/vendor/codeception/codeception/src/Codeception/Lib/Di.php:127
/var/www/vendor/codeception/codeception/src/Codeception/Test/Cest.php:138
/var/www/vendor/codeception/codeception/src/Codeception/Test/Cest.php:150
/var/www/vendor/codeception/codeception/src/Codeception/Test/Cest.php:82
/var/www/vendor/codeception/codeception/src/Codeception/Test/Test.php:88
/var/www/vendor/phpunit/phpunit/src/Framework/TestSuite.php:733
/var/www/vendor/codeception/phpunit-wrapper/src/Runner.php:110
/var/www/vendor/codeception/codeception/src/Codeception/SuiteManager.php:158
/var/www/vendor/codeception/codeception/src/Codeception/Codecept.php:193
/var/www/vendor/codeception/codeception/src/Codeception/Codecept.php:160
/var/www/vendor/codeception/codeception/src/Codeception/Command/Run.php:391
/var/www/vendor/symfony/console/Command/Command.php:245
/var/www/vendor/symfony/console/Application.php:835
/var/www/vendor/symfony/console/Application.php:185
/var/www/vendor/symfony/console/Application.php:117
/var/www/vendor/codeception/codeception/src/Codeception/Application.php:107
/var/www/vendor/codeception/codeception/codecept:43

Artifacts:

Html: /var/www/tests/_output/EmailCest.trySendActivationEmail.fail.html
Response: /var/www/tests/_output/EmailCest.trySendActivationEmail.fail.html

---------
2) EmailCest: Try send notification email
 Test  tests/functional/EmailCest.php:trySendNotificationEmail
 Step  See in current url "/user/welcome"
 Fail  Failed asserting that '/user/create' contains "/user/welcome".

Scenario Steps:

 14. $I->seeInCurrentUrl("/user/welcome") at tests/functional/EmailCest.php:124
 13. $I->click("Register") at tests/functional/EmailCest.php:122
 12. $I->fillField({"id":"User_verifyCode"},"boom") at tests/functional/EmailCest.php:121
 11. $I->checkOption("#User_terms") at tests/functional/EmailCest.php:120
 10. $I->checkOption("#User_newsletter") at tests/functional/EmailCest.php:119
 9. $I->selectOption("form select[id=User_preferred_link]","NCBI") at tests/functional/EmailCest.php:118

/var/www/vendor/phpunit/phpunit/src/Framework/Constraint.php:111
/var/www/vendor/phpunit/phpunit/src/Framework/Constraint.php:56
/var/www/vendor/phpunit/phpunit/src/Framework/Assert.php:2060
/var/www/vendor/phpunit/phpunit/src/Framework/Assert.php:146
/var/www/vendor/codeception/phpunit-wrapper/src/TestCase.php:47
/var/www/vendor/codeception/codeception/src/Codeception/Util/Shared/Asserts.php:449
/var/www/vendor/codeception/codeception/src/Codeception/Lib/InnerBrowser.php:562
/var/www/vendor/codeception/codeception/src/Codeception/Step.php:266
/var/www/vendor/codeception/codeception/src/Codeception/Scenario.php:72
/var/www/tests/_support/_generated/FunctionalTesterActions.php:1863
/var/www/tests/functional/EmailCest.php:124
/var/www/vendor/codeception/codeception/src/Codeception/Lib/Di.php:127
/var/www/vendor/codeception/codeception/src/Codeception/Test/Cest.php:138
/var/www/vendor/codeception/codeception/src/Codeception/Test/Cest.php:150
/var/www/vendor/codeception/codeception/src/Codeception/Test/Cest.php:82
/var/www/vendor/codeception/codeception/src/Codeception/Test/Test.php:88
/var/www/vendor/phpunit/phpunit/src/Framework/TestSuite.php:733
/var/www/vendor/codeception/phpunit-wrapper/src/Runner.php:110
/var/www/vendor/codeception/codeception/src/Codeception/SuiteManager.php:158
/var/www/vendor/codeception/codeception/src/Codeception/Codecept.php:193
/var/www/vendor/codeception/codeception/src/Codeception/Codecept.php:160
/var/www/vendor/codeception/codeception/src/Codeception/Command/Run.php:391
/var/www/vendor/symfony/console/Command/Command.php:245
/var/www/vendor/symfony/console/Application.php:835
/var/www/vendor/symfony/console/Application.php:185
/var/www/vendor/symfony/console/Application.php:117
/var/www/vendor/codeception/codeception/src/Codeception/Application.php:107
/var/www/vendor/codeception/codeception/codecept:43

Artifacts:

Html: /var/www/tests/_output/EmailCest.trySendNotificationEmail.fail.html
Response: /var/www/tests/_output/EmailCest.trySendNotificationEmail.fail.html
```

If I open the html artefacts, I get this error on email field:

```

Email * swordmaster@mailinator.com
Email "swordmaster@mailinator.com" has already been taken.

````

It seems like the data entered in the database during the execution of those tests are not reset correctly maybe?

Maybe you want to add a ``bootstrap:`` clause to the ``tests/functional.suite.yml`` file like we did for ``tests/acceptance.suite.yml`` and ``tests/unit.suite.yml``

File: protected/components/PasswordResetTokenUserIdentity.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        {
+            $this->errorCode = self::ERROR_SELECTOR_NOT_ASSOCIATED_WITH_A_USER;
+        } 
+        else  // User found
+        {
+            // Check re-calculated hash from verifier matches hash in reset_password_request database table
+            $signingKey = Yii::app()->params['signing_key'];
+            $verifierFromURL = substr($this->urlToken, 20, 20);
+            $hashedTokenFromURLVerifier = Yii::app()->CryptoService->getHashedToken($signingKey, $verifierFromURL);
+            if($hashedTokenFromURLVerifier == $resetPasswordRequest->hashed_token)
```
Comment (rija) [inline protected/components/PasswordResetTokenUserIdentity.php:52]: we really should use the stricter ``===`` operator instead of the ``==`` one

**Review by rija — CHANGES_REQUESTED** [review]

**Comment by pli888** [comment]
>So it's probably a good idea to update CryptoService::getRandomAlphNumStr() to use the Yii2-provided method

Yii2's `generateRandomString()` is now used in `CryptoService.php`.

>The source code of the Yii2 method is in vendor/yiisoft/yii2/base/Security.php, may be there are other useful methods in there?

I've looked at the code in `vendor/yiisoft/yii2/base/Security.php` and it looks like there is useful code but I am thinking to leave this for future work in order to get this PR merged for the demo?

**Comment by pli888** [comment]
>It seems like the data entered in the database during the execution of those tests are not reset correctly maybe?

Yes, data in the database needs to be reset in order for the tests to be re-run. A bootstrap clause has been added to reset the database for functional tests in `functional.suite.yml`.

File: protected/models/ResetPasswordRequest.php [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        );
+    }
+
+    /**
+     * The requested_at and expires_at timestamps are created if they have been
+     * provided before saving a record into the reset_password_request table.
+     * 
+     * @return bool
+     */
+    public function beforeSave()
```
[the remark this answers is not in the document]
Comment (pli888) [reply protected/models/ResetPasswordRequest.php:80]: `parent::beforeSave()` is now the first instruction and is tested too before the remaining instructions are executed.

File: tests/unit/CryptoTest.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    /**
+     * Verifies hashing algorithm produces expected results
+     */
+    public function testGetHashedToken()
+    {
+        $hashValue = "PMzM5S0Q2z6Eg+rc7D/3SkB3c4RQ1PNJzRtQ/bN0/nU=";
+        $data = "hobbit";
+        $key = "1234567890";
```
[the remark this answers is not in the document]
Comment (pli888) [reply tests/unit/CryptoTest.php:26]: Information about the HASH_SECRET_KEY has been added into docs/variables.md.

File: ops/configuration/yii-conf/local.php.dist [hunk 4]
```diff
@@ -2,6 +2,9 @@
 
 return array(
     'components' => array(
+        'CryptoService' => array(
```
[the remark this answers is not in the document]
Comment (pli888) [reply ops/configuration/yii-conf/local.php.dist]: The name of the component now begins with a lower case 'c'.

File: protected/components/PasswordResetTokenUserIdentity.php (hunk 1 above)
Comment (pli888) [reply protected/components/PasswordResetTokenUserIdentity.php]: This is done now.

File: protected/components/PasswordResetTokenUserIdentity.php [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+        if ($user === null)  // User not found
+        {
+            $this->errorCode = self::ERROR_SELECTOR_NOT_ASSOCIATED_WITH_A_USER;
+        } 
+        else  // User found
+        {
+            // Check re-calculated hash from verifier matches hash in reset_password_request database table
+            $signingKey = Yii::app()->params['signing_key'];
+            $verifierFromURL = substr($this->urlToken, 20, 20);
```
[the remark this answers is not in the document]
Comment (pli888) [reply protected/components/PasswordResetTokenUserIdentity.php]: A constant called `RANDOM_ALPHA_NUMERIC_STRING_LENGTH` has been defined to provide the length of the random alpha numeric string in `CryptoService`.

**Review by rija — APPROVED** [review]
Hi @pli888,

all tests are passing on my local dev environment and on CI.

After inserting in the database a new user with email address under my control on my local staging environment, I have managed to use the functionality described below:

* [x] Happy path (successfully reset and changed my password)
* [x] Clicking on the reset link again after having changed password resulted in error "Your password reset token is invalid. Please request another."
* [x] Waiting more than one hour before clicking the reset link resulted in error "Your password reset token is invalid. Please request another."
* [x] Tampering with the token in the reset link resulted in error "Your password reset token is invalid. Please request another."


The code looks good now, so I'm happy to approve this PR

**Comment by kencho51** [comment]
Hi @pli888,

In the reset page, would it be better if the buttons `Cancel` and Save` are aligned?
 
image

**Comment by kencho51** [comment]
Hi @pli888,

In my staging gigadb, using chrome browser, the button's position in the forgot page and reset page are not positioned well. but the problems are not observed in safari.
 
 
image


image

**Comment by pli888** [comment]
Hi @kencho51 @rija 

>In my dev and staging gigadb website, using chrome browser, the button's position in the forgot page and reset page are not positioned well. but the problems are not observed in safari.

The problems with the alignment of the buttons in the reset password feature are due to changes to `less/current/modules/forms.less` earlier this month when trying to fix the layout of the HTML elements on the dataset admin page. If we look at other web pages with forms, they also have layout problems, e.g. [/datasetSubmission/create1](https://staging.gigadb.org/datasetSubmission/create1) in the current dataset submission wizard:

datasetSubmissionCreate

In hindsight, I should have paid attention to these pages with web forms when I did the CSS work. I suggest we revert the latest changes in the `less` files and re-add issue [Admin pages - Misalignment of Update Dataset page #926](#926) into the sprint backlog. We then look into how to re-create the original CSS files more faithfully from the less files in order to fix the layout problems with the web page forms.

**Comment by pli888** [comment]
@kencho51 The problem with `testGetRandomAlphaNumStr` in the CryptoTest unit test is due to the presence of underscore and dash characters in the random string. These are now removed before `ctype_alnum()` is used to test the random alphanumeric string successfully.

**Comment by kencho51** [comment]
Hi @pli888,

> @kencho51 The problem with `testGetRandomAlphaNumStr` in the CryptoTest unit test is due to the presence of underscore and dash characters in the random string. These are now removed before `ctype_alnum()` is used to test the random alphanumeric string successfully.

I have re-run the CryptoTest unit test and it is passing. But if the `$randomStr` is generated with `_` and/or `-`, after `str_replace`, the length of  `$randomStr` will be less than `20` char. So, these two assertions seems contradicted in this case :
```
$this->assertTrue(strlen($randomStr) == 20, "Random string does not contain 20 characters");
$this->assertTrue(ctype_alnum(str_replace($valid_chars, "", $randomStr)), "Random string is not alphanumeric");
```

Also the `CryptoService::getRandomAlphaNumStr()` would still produce random string containing underscore and dash?

In fact, my question would be is it ok to have a token with `_` and `-`?  If yes, the `ctype_alnum` assertion could be removed. If no, a new implementation, [for example](https://tutorialio.com/generate-a-random-alphanumeric-string-in-php/), of generating a random alpha numeric string would be needed as `CryptoService::getRandomAlphaNumStr()` would produce string with `_` and `-` by default.

**Comment by pli888** [comment]
@kencho51 

>$this->assertTrue(strlen($randomStr) == 20, "Random string does not contain 20 characters");
$this->assertTrue(ctype_alnum(str_replace($valid_chars, "", $randomStr)), "Random string is not alphanumeric");

These 2 assertions work together to test the random string generated by `CryptoService::getRandomAlphaNumStr()`.

The first assertion checks that the random string from `CryptoService` contains 20 characters. I do not want to check the random string length = 20 in the second assertion because this has already been tested in the first assertion, I just want to check it is an alphanumeric string.

It is convenient to use the `ctype_alnum` function to check the random string only contains alphanumeric characters - the problem is that you cannot use the `ctype_alnum` function with underscores and dashes. However, `CryptoService::getRandomAlphaNumStr()` uses the `generateRandomString()` function from Yii security package as a standard way of generating random strings in the `gigadb-website` codebase and it includes underscores and dashes in its random strings. Because I know this, I know that these need to be removed before the assertion is done with `ctype_alnum`.

**Comment by pli888** [comment]
@rija @kencho51 

The `getRandomAlphaNumStr()` function in `CryptoService` has been renamed to `getRandomString()` since it also uses underscore and dash characters in the random strings it creates. The acceptance and integration tests have also been updated to use test selector and verifier values that contain underscores and dashes to ensure that the reset password feature can work with these characters.

Finally, the `forms.less` file has been reverted back to what it looked like when the initial CSS less work was done. This fixes the alignment of the Reset, Cancel and Submit buttons.

**Review by kencho51 — APPROVED** [review]
Hi @pli888,

The tests are passing locally and in my gitlab pipeline, and the reset password functionality works as expected in my dev and staging gigadb website.

I could see the reset password link consists of two random strings(selector + verifier), and will get `Error 403 You are not authorized to perform this action.` if click the reset link second time.

I could confirm that  I will get this error `Your password reset token is invalid. Please request another.`  if click the reset password link after 1 hour it was sent out.

The code looks good to me too, so I would like approve this PR.

**Pull request merged**